frontend-react/docs/data-fetching-patterns.md

```diff
@@ -0,0 +1,312 @@
+
+## Background
+
+Over its lifetime, the ReportStream frontend has been through a number of revisions in regards to its approach to data fetching.
+
+The first concerted approach to standardization revolved around the [rest-hooks](https://resthooks.io/) library. Rest-hooks took a class-based approach to defining an API interface and provided a lot of useful functionality, but in the end, in addition to being somewhat complex to use and test, and underdocumented, was not flexible enough to handle our needs working with a somewhat fickle and non-RESTful API. We dropped the library and built our own layer(s).
+
+In our two attempts and building our own system without the help of an external library largely faltered around the issues of maintenance and reliability, particularly in regards to state management, and flexibility for handling requests of different types. A decision was made to rethink our approach holistically and consider bringing in a new third party library.
+
+As documented [here](https://cdc.sharepoint.com/:w:/r/teams/USDSatCDC/Shared%20Documents/PRIME%20ReportStream/Team%20Documents/Experience%20Team/Engineering/ReportStream%20Frontend%20Data%20Fetching%20Design%20Research.docx?d=wfe7ad07f71cc4845acc66ec2a1022226&csf=1&web=1&e=lfTMeB), a decision was made to implement [react-query](), and build a new data fetching system around it that would provide the flexibility and ease of use lacking in rest-hooks, while also being easy to maintain and feature rich, unlike our home grown solutions.
+
+## Component Level
+
+All components that require data from the API, or that push data to the API, should do so through simple invocation of custom hooks that will provide:
+
+-   data, when available
+-   errors, when available
+-   loading state
+
+In general, fetch logic shouldn't live in components, and all a component should need to do is invoke a hook!
+
+### Typing
+
+Custom hooks can be generically typed in order to provide various types of return values, or to take multiple types of arguments. In the example below, the `useValueSetsTable` hook can return either `ValueSet[]` or `ValueSetRow[]`, so this type is specified in the generic for each hook invocation depending on what the component expects.
+
+### Example
+
+A simplified example taken from [ValueSetsIndex.tsx](https://github.com/CDCgov/prime-reportstream/blob/58599e355ad2a0b357005190c185a8b8497f0b81/frontend-react/src/pages/admin/value-set-editor/ValueSetsIndex.tsx)
+
+```typescript
+import { useValueSetsTable } from "../../../hooks/UseValueSets";
+const ValueSetsTable = () => {
+    const { valueSetArray, isLoading, error } = useValueSetsTable<ValueSet[]>(
+        LookupTables.VALUE_SET
+    );
+    if (isLoading) {
+    	// may use suspense, but if not, you can render a loading state here
```

Comment:
This seems like it encourages the pattern we chose not to implement for loading and error state. Maybe move the example use of `isLoading` and `error` into a separate section?

Reply:
I left the comments here as a note that these may not be the particular patterns for handling loading and error states, but more a demonstration of what the hook may return. I feel like we should revisit this once we set out the pattern for loading and error handling, and maybe write up a full separate document on that if it's warranted